server/channels/store/sqlstore/channel_store.go

```diff
@@ -1922,6 +1922,93 @@ func (s SqlChannelStore) UpdateMemberNotifyProps(channelID, userID string, props
 	return dbMember.ToModel(), err
 }
 
+func (s SqlChannelStore) UpdateMultipleMembersNotifyProps(members []*model.ChannelMember) ([]*model.ChannelMember, error) {
```

Comment:
This is the same as serial calls to `UpdateMemberNotifyProps` wrapped in a transation - which will create a lot of contention in that table, because there is a potentially long transaction with unbounded length, and there are multiple SQL queries with both UPDATE and SELECT.

My first concern is that the input is unbounded.
1. We should have a limit to the number of members from the app layer.
2. I know we always return the updated channel member value, but I want to take a pause here and try to understand if the client actually has a necessity to use the updated value. Because the SQL query for the `SELECT` joins multiple tables, and if you have to run this 100 times in a single tx then it's not exactly very good. I'd suggest to check exactly what values do the client need (if any), and send only those.
3. We really need to make multiple UPDATEs in a single query. I've found this: https://stackoverflow.com/a/70803859/1027058 to be an interesting approach. It's okay to optimize this only for Postgres.

Reply:
1. Makes sense. Is there a standard limit I should use?
2. Yeah, I can remove that. I actually didn't return the updated notify props originally, and I only added that to more closely match the existing API, but I don't think that needs to be done really. I also tried merging that into a single `SELECT ... WITH (userId, channelId) in ((user1, channel1), (user2, channel2), ...)`, but I don't know if that would actually help performance at all.
3. That looks interesting. It'd definitely be nice to do that without multiple queries, so I can take a look. Alternatively, if we're really concerned about performance here, I could replace this with a less general API that patches multiple users' channel members with the same notify props since our use case would be to set them all to `{mark_unread: 'mention', msteams_automute: 'true'}` without modifying the other notify props. Then, we could likely write the query as `UPDATE channelMembers SET notifyprops || ?::jsonb WHERE (userId, channelId) in ((user1, channel1), (user2, channel2), ...)`

Reply:
1. 200 has been our usual max default: https://github.com/mattermost/mattermost/blob/cd8796bf0b22239db286464a4282b515ca535af2/server/channels/web/params.go#L20
2. Ok then let's remove it.
3. > Alternatively, if we're really concerned about performance here, I could replace this with a less general API that patches multiple users' channel members with the same notify props 

    Great, that would be perfect! Let's do this.

Reply:
2. I found out that we still need this actually because we send a generic `ChannelMemberUpdated` WebSocket event for notify props changes. We could add a more specific event for this in the future, but for backwards compatibility, I'm not going to change that for now